krabs/krabs/schema_locator.hpp

```diff
@@ -32,28 +32,84 @@ namespace krabs {
     struct schema_key
     {
         guid      provider;
+
+        // Using a string_view for name so that keys can be constructed from
+        // EVENT_RECORD pointers without allocation. If the key instance is
+        // added to the cache, 'internalize_name' must be called first so that
+        // the name points to owned memory and doesn't dangle.
+        // Only events logged with the TraceLogger API will have a name set
+        // in the key because it's available as part of the EVENT_RECORD.
+        // Other events are uniquely distinguished by their event Id.
+        std::string_view name;
+
         uint16_t  id;
-        uint8_t   opcode;
         uint8_t   version;
+        uint8_t   opcode;
         uint8_t   level;
+        uint64_t  keyword;
+
+    private:
+        // See note on 'name', this is only set when internalized and
+        // provides memory ownership for the string_view.
+        std::unique_ptr<std::string> backing_name;
 
-        schema_key(const EVENT_RECORD &record)
+    public:
+        schema_key(const EVENT_RECORD &record, std::string_view name)
             : provider(record.EventHeader.ProviderId)
+            , name(name)
             , id(record.EventHeader.EventDescriptor.Id)
+            , version(record.EventHeader.EventDescriptor.Version)
             , opcode(record.EventHeader.EventDescriptor.Opcode)
             , level(record.EventHeader.EventDescriptor.Level)
-            , version(record.EventHeader.EventDescriptor.Version) { }
+            , keyword(record.EventHeader.EventDescriptor.Keyword) { }
```

Comment:
Now that the name is a part of the key, do we need the keyword? Or did we add the keyword because we want to add all attributes of the event descriptor to a part of the schema key?